### PR TITLE
Add reservedInternalRange fields to the google_compute_subnetwork resource

### PR DESCRIPTION
**Pull request by daanheikens**
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/17881

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added fields `reserved_internal_range` and `secondary_ip_ranges[].reserved_internal_range` to `google_compute_subnetwork` resource
```

### CODE REVIEW

File: mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+resource "google_network_connectivity_internal_range" "reserved_range" {
+  name              = "reserved-range"
+  network           = google_compute_network.default.id
+  usage             = "FOR_VPC"
+  peering           = "FOR_SELF"
+  prefix_length     = 24
+  target_cidr_range = [
+    "10.0.0.0/8"
+  ]
+}
```
Comment (bschaatsbergen) [inline mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb:22]: add a newline

File: mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  network                 = google_compute_network.default.id
+  reserved_internal_range = google_network_connectivity_internal_range.reserved_range.id
+}
+
+resource "google_compute_network" "default" {
+  name                    = "<%= ctx[:vars]['network_name'] %>"
+  auto_create_subnetworks = false
+}
+
+resource "google_network_connectivity_internal_range" "reserved_range" {
```
Comment (bschaatsbergen) [inline mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb:13]:
```suggestion
resource "google_network_connectivity_internal_range" "reserved" {
```

File: mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range" {
+  name                    = "<%= ctx[:vars]['subnetwork_name'] %>"
+  region                  = "us-central1"
+  network                 = google_compute_network.default.id
+  reserved_internal_range = google_network_connectivity_internal_range.reserved_range.id
```
Comment (bschaatsbergen) [inline mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb:5]:
```suggestion
  reserved_internal_range = google_network_connectivity_internal_range.reserved.id
```

File: mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+resource "google_compute_network" "default" {
+  name                    = "<%= ctx[:vars]['network_name'] %>"
+  auto_create_subnetworks = false
+}
+
+resource "google_network_connectivity_internal_range" "reserved_range" {
+  name              = "reserved-range"
```
Comment (bschaatsbergen) [inline mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb:14]:
```suggestion
  name              = "reserved"
```

File: mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb (hunk 3 above)
Comment (bschaatsbergen) [reply mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb:5]: Interesting, the API spec suggests specifying the URL, but using the ID works too. Maybe we can update the parameter description? What do you think?

**Review by bschaatsbergen — APPROVED** [review]
LGTM 🚀

**Comment by daanheikens** [comment]
Hi @trodge, happy to receive your review on this PR. Thanks!

File: mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range" {
+  name                    = "<%= ctx[:vars]['subnetwork_name'] %>"
```
Comment (trodge) [inline mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb:2]: You'll need to add `provider = google-beta` to these resources.

Comment (daanheikens) [reply mmv1/templates/terraform/examples/subnetwork_reserved_internal_range.tf.erb:3]: @trodge added!

**Review by bschaatsbergen — APPROVED** [review]
LGTM :rock

**Comment by trodge** [comment]
The test is failing with this error:
```
2024-04-30T18:59:52.071Z [ERROR] sdk.helper_resource: Unexpected error: test_working_directory=/tmp/plugintest168036712 test_step_number=1
  error=
  | Error running pre-apply refresh: exit status 1
  | 
  | Error: Missing required argument
  | 
  |   on terraform_plugin_test.tf line 2, in resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range":
  |    2: resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range" {
  | 
  | The argument "ip_cidr_range" is required, but no definition was found.
   test_name=TestAccComputeSubnetwork_subnetworkReservedInternalRangeExample test_terraform_path=/bin/terraform
```

**Comment by daanheikens** [comment]
> The test is failing with this error:
> 
> ```
> 2024-04-30T18:59:52.071Z [ERROR] sdk.helper_resource: Unexpected error: test_working_directory=/tmp/plugintest168036712 test_step_number=1
>   error=
>   | Error running pre-apply refresh: exit status 1
>   | 
>   | Error: Missing required argument
>   | 
>   |   on terraform_plugin_test.tf line 2, in resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range":
>   |    2: resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range" {
>   | 
>   | The argument "ip_cidr_range" is required, but no definition was found.
>    test_name=TestAccComputeSubnetwork_subnetworkReservedInternalRangeExample test_terraform_path=/bin/terraform
> ```

Hey @trodge, It appears that `ipCidrRange` becomes optional once `reserved_internal_range` is defined. Currently, it's set as required. What is the recommended approach here? I see two options:

1. Making `ipCidrRange` in both primary and secondary ranges optional and document the use case when it's required
2. Creating custom code to populate those fields once `ipCidrRange` is empty (in either primary/secondary) and `reserved_internal_range` is set

I Just tested option 1, setting `ipCidrRange` to optional and `default_from_api` to true, then test it again and the test passes.

Happy to hear your thoughts.

**Comment by trodge** [comment]
> > The test is failing with this error:
> > ```
> > 2024-04-30T18:59:52.071Z [ERROR] sdk.helper_resource: Unexpected error: test_working_directory=/tmp/plugintest168036712 test_step_number=1
> >   error=
> >   | Error running pre-apply refresh: exit status 1
> >   | 
> >   | Error: Missing required argument
> >   | 
> >   |   on terraform_plugin_test.tf line 2, in resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range":
> >   |    2: resource "google_compute_subnetwork" "subnetwork_with_reserved_internal_range" {
> >   | 
> >   | The argument "ip_cidr_range" is required, but no definition was found.
> >    test_name=TestAccComputeSubnetwork_subnetworkReservedInternalRangeExample test_terraform_path=/bin/terraform
> > ```
> 
> Hey @trodge, It appears that `ipCidrRange` becomes optional once `reserved_internal_range` is defined. Currently, it's set as required. What is the recommended approach here? I see two options:
> 
> 1. Making `ipCidrRange` in both primary and secondary ranges optional and document the use case when it's required
> 2. Creating custom code to populate those fields once `ipCidrRange` is empty (in either primary/secondary) and `reserved_internal_range` is set
> 
> I Just tested option 1, setting `ipCidrRange` to optional and `default_from_api` to true, then test it again and the test passes.
> 
> Happy to hear your thoughts.

Making the field optional should be fine.

**Review by trodge — APPROVED** [review]

**Comment by daanheikens** [comment]
> It looks like this needed to be reverted due to a breaking change. We will probably need to make this change in the next major release.

Ah pity! Anyway, happy to help. How should we proceed from here @trodge?

**Comment by daanheikens** [comment]
@trodge i saw the announcement for Terraform v6. Is this change coming to v6? I would really be helped having this feature. Is there maybe a way to integrate internal ranges on just the primary range?

**Pull request merged**